File: lang/go/cxlib/table.go
// CX Go binding — Public Table API per ADR 0018 D1.
//
// Implements the 17-member canonical Table API surface against the
// V core's :table blocks via the C ABI. Per ADR 0018 §D2 per-binding
// naming: Go uses PascalCase methods + lowerCamelCase struct fields
// to match host conventions.

package cxlib

import (
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"
)

// Table is an immutable handle over a single :table block. Per
// ADR 0018 §D3 tables are immutable values; modification returns a
// new Table. Cells admit any ADR 0017 §D5 Item kind: scalars +
// Array (Go []any) + Map (Go map[string]any).
type Table struct {
	cols  []string
	types []string
	rows  [][]any
}

// ── Construction ─────────────────────────────────────────────────────────────

// TableFromCx parses CX source and returns the first :table block
// found, walking nested structures. Returns an error if no :table
// block is present.
func TableFromCx(src string) (*Table, error) {
	tables, err := TablesFromCx(src)
	if err != nil {
		return nil, err
	}
	if len(tables) == 0 {
		return nil, errors.New("cxlib: no :table block found in source")
	}
	return tables[0], nil
}

// TablesFromCx returns every :table block in the source in
// document order. Walks the data_bin decode recursively.
func TablesFromCx(src string) ([]*Table, error) {
	framed, err := ToDataBin(src)
	if err != nil {
		return nil, err
	}
	decoded, err := FromDataBinValue(framed)
	if err != nil {
		return nil, err
	}
	var tables []*Table
	collectTables(decoded, &tables)
	return tables, nil
}

// NewTable constructs a Table directly with 4-invariant validation
// per ADR 0018 §D7: len(cols)==len(types), unique col names, row
// width matches col count, (cell type compatibility is checked at
// access time since Go interfaces are duck-typed).
func NewTable(cols []string, types []string, rows [][]any) (*Table, error) {
	if len(cols) != len(types) {
		return nil, fmt.Errorf("cxlib: len(cols)=%d != len(types)=%d",
			len(cols), len(types))
	}
	seen := make(map[string]struct{}, len(cols))
	for _, c := range cols {
		if _, ok := seen[c]; ok {
			return nil, fmt.Errorf("cxlib: duplicate column name %q", c)
		}
		seen[c] = struct{}{}
	}
	for rowIdx, row := range rows {
		if len(row) != len(cols) {
			return nil, fmt.Errorf(
				"cxlib: row %d has %d cells; expected %d",
				rowIdx, len(row), len(cols),
			)
		}
	}
	colsCopy := make([]string, len(cols))
	copy(colsCopy, cols)
	typesCopy := make([]string, len(types))
	copy(typesCopy, types)
	rowsCopy := make([][]any, len(rows))
	for i, row := range rows {
		r := make([]any, len(row))
		copy(r, row)
		rowsCopy[i] = r
	}
	return &Table{cols: colsCopy, types: typesCopy, rows: rowsCopy}, nil
}

// ── Properties (4) ───────────────────────────────────────────────────────────

func (t *Table) Cols() []string {
	out := make([]string, len(t.cols))
	copy(out, t.cols)
	return out
}

func (t *Table) Types() []string {
	out := make([]string, len(t.types))
	copy(out, t.types)
	return out
}

func (t *Table) RowCount() int { return len(t.rows) }
func (t *Table) ColCount() int { return len(t.cols) }

// ── Access (9) ───────────────────────────────────────────────────────────────

// Row returns the row at index i as an ordered map.
func (t *Table) Row(i int) (map[string]any, error) {
	if i < 0 || i >= len(t.rows) {
		return nil, fmt.Errorf("cxlib: row index %d out of bounds [0, %d)",
			i, len(t.rows))
	}
	out := make(map[string]any, len(t.cols))
	for c, name := range t.cols {
		out[name] = t.rows[i][c]
	}
	return out, nil
}

// Column returns all values in the named column.
func (t *Table) Column(name string) ([]any, error) {
	idx := -1
	for i, c := range t.cols {
		if c == name {
			idx = i
			break
		}
	}
	if idx < 0 {
		return nil, fmt.Errorf("cxlib: unknown column %q", name)
	}
	return t.ColAt(idx)
}

func (t *Table) ColAt(i int) ([]any, error) {
	if i < 0 || i >= len(t.cols) {
		return nil, fmt.Errorf("cxlib: column index %d out of bounds [0, %d)",
			i, len(t.cols))
	}
	out := make([]any, len(t.rows))
	for ri, row := range t.rows {
		out[ri] = row[i]
	}
	return out, nil
}

func (t *Table) Cell(r, c int) (any, error) {
	if r < 0 || r >= len(t.rows) {
		return nil, fmt.Errorf("cxlib: row index %d out of bounds [0, %d)",
			r, len(t.rows))
	}
	if c < 0 || c >= len(t.cols) {
		return nil, fmt.Errorf("cxlib: column index %d out of bounds [0, %d)",
			c, len(t.cols))
	}
	return t.rows[r][c], nil
}

func (t *Table) CellByName(r int, name string) (any, error) {
	for c, colName := range t.cols {
		if colName == name {
			return t.Cell(r, c)
		}
	}
	return nil, fmt.Errorf("cxlib: unknown column %q", name)
}

func (t *Table) Slice(start, end int) (*Table, error) {
	if start < 0 || start > len(t.rows) {
		return nil, fmt.Errorf("cxlib: slice start %d out of bounds", start)
	}
	if end < start || end > len(t.rows) {
		return nil, fmt.Errorf("cxlib: slice end %d out of bounds (start=%d)",
			end, start)
	}
	newRows := make([][]any, end-start)
	for i, row := range t.rows[start:end] {
		r := make([]any, len(row))
		copy(r, row)
		newRows[i] = r
	}
	return &Table{cols: t.cols, types: t.types, rows: newRows}, nil
}

func (t *Table) Head(n int) *Table {
	if n < 0 {
		n = 0
	}
	if n > len(t.rows) {
		n = len(t.rows)
	}
	out, _ := t.Slice(0, n)
	return out
}

func (t *Table) Tail(n int) *Table {
	start := len(t.rows) - n
	if start < 0 {
		start = 0
	}
	out, _ := t.Slice(start, len(t.rows))
	return out
}

// Select returns a new Table with only the named columns in the given
// order. Per ADR 0018 §4 canonical name; Go allows `select` since
// it's only a keyword in channel syntax (not a function name).
func (t *Table) Select(names []string) (*Table, error) {
	newCols := make([]string, 0, len(names))
	newTypes := make([]string, 0, len(names))
	indices := make([]int, 0, len(names))
	for _, name := range names {
		idx := -1
		for i, c := range t.cols {
			if c == name {
				idx = i
				break
			}
		}
		if idx < 0 {
			return nil, fmt.Errorf("cxlib: unknown column %q", name)
		}
		newCols = append(newCols, t.cols[idx])
		newTypes = append(newTypes, t.types[idx])
		indices = append(indices, idx)
	}
	newRows := make([][]any, len(t.rows))
	for ri, row := range t.rows {
		r := make([]any, len(indices))
		for j, idx := range indices {
			r[j] = row[idx]
		}
		newRows[ri] = r
	}
	return &Table{cols: newCols, types: newTypes, rows: newRows}, nil
}

// ── Iteration (2) ────────────────────────────────────────────────────────────

// Iter returns a channel of rows (each as an ordered map). Per Go
// convention; ranges close the channel when iteration completes.
func (t *Table) Iter() <-chan map[string]any {
	ch := make(chan map[string]any, 1)
	go func() {
		defer close(ch)
		for i := range t.rows {
			row, _ := t.Row(i)
			ch <- row
		}
	}()
	return ch
}

// ColView is the (name, type_name, values) triple yielded by IterCols.
type ColView struct {
	Name     string
	TypeName string
	Values   []any
}

func (t *Table) IterCols() <-chan ColView {
	ch := make(chan ColView, 1)
	go func() {
		defer close(ch)
		for i, name := range t.cols {
			vals, _ := t.ColAt(i)
			ch <- ColView{Name: name, TypeName: t.types[i], Values: vals}
		}
	}()
	return ch
}

// ── Conversion (5) ───────────────────────────────────────────────────────────

func (t *Table) ToCx() string {
	headerParts := make([]string, len(t.cols))
	for i, name := range t.cols {
		if t.types[i] == "" {
			headerParts[i] = name
		} else {
			headerParts[i] = name + ":" + t.types[i]
		}
	}
	var b strings.Builder
	b.WriteString("[_ :table[")
	b.WriteString(strings.Join(headerParts, " "))
	b.WriteString("]\n")
	for _, row := range t.rows {
		b.WriteString("  ")
		for j, v := range row {
			if j > 0 {
				b.WriteString(" ")
			}
			b.WriteString(formatCxCell(v))
		}
		b.WriteString("\n")
	}
	b.WriteString("]\n")
	return b.String()
}

func (t *Table) ToCSV(delim byte) string {
	var b strings.Builder
	for i, c := range t.cols {
		if i > 0 {
			b.WriteByte(delim)
		}
		b.WriteString(c)
	}
	b.WriteString("\r\n")
	for _, row := range t.rows {
		for j, v := range row {
			if j > 0 {
				b.WriteByte(delim)
			}
			b.WriteString(formatCsvCell(v, delim))
		}
		b.WriteString("\r\n")
	}
	return b.String()
}

func (t *Table) ToJSON() (string, error) {
	rows := make([]map[string]any, len(t.rows))
	for i := range t.rows {
		row, _ := t.Row(i)
		rows[i] = row
	}
	out, err := json.Marshal(rows)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

func (t *Table) ToDataBin() ([]byte, error) {
	return ToDataBin(t.ToCx())
}

func (t *Table) ToDictList() []map[string]any {
	out := make([]map[string]any, len(t.rows))
	for i := range t.rows {
		out[i], _ = t.Row(i)
	}
	return out
}

// ── Equality ─────────────────────────────────────────────────────────────────

func (t *Table) Equal(other *Table) bool {
	if other == nil {
		return false
	}
	if len(t.cols) != len(other.cols) || len(t.rows) != len(other.rows) {
		return false
	}
	for i, c := range t.cols {
		if c != other.cols[i] || t.types[i] != other.types[i] {
			return false
		}
	}
	for ri, row := range t.rows {
		if len(row) != len(other.rows[ri]) {
			return false
		}
		for j, v := range row {
			if !cellEqual(v, other.rows[ri][j]) {
				return false
			}
		}
	}
	return true
}

func cellEqual(a, b any) bool {
	// Slice / map equality requires deep compare; use JSON round-trip
	// for a simple, correct-for-our-types implementation.
	ja, _ := json.Marshal(a)
	jb, _ := json.Marshal(b)
	return string(ja) == string(jb)
}

// ── Internal: walk a decoded data_bin payload to find tables ─────────────────

func collectTables(value any, out *[]*Table) {
	switch v := value.(type) {
	case map[string]any:
		// Walk children
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			collectTables(v[k], out)
		}
	case []any:
		if looksLikeTable(v) {
			first := v[0].(map[string]any)
			cols := make([]string, 0, len(first))
			for k := range first {
				cols = append(cols, k)
			}
			sort.Strings(cols)
			types := make([]string, len(cols))
			rows := make([][]any, len(v))
			for ri, item := range v {
				r := make([]any, len(cols))
				m := item.(map[string]any)
				for ci, c := range cols {
					r[ci] = m[c]
				}
				rows[ri] = r
			}
			*out = append(*out, &Table{cols: cols, types: types, rows: rows})
		} else {
			for _, child := range v {
				collectTables(child, out)
			}
		}
	}
}

func looksLikeTable(v []any) bool {
	if len(v) == 0 {
		return false
	}
	first, ok := v[0].(map[string]any)
	if !ok || len(first) == 0 {
		return false
	}
	keys := make(map[string]struct{}, len(first))
	for k := range first {
		keys[k] = struct{}{}
	}
	for _, item := range v {
		m, ok := item.(map[string]any)
		if !ok || len(m) != len(keys) {
			return false
		}
		for k := range m {
			if _, present := keys[k]; !present {
				return false
			}
		}
	}
	return true
}

// ── Internal: cell formatters ────────────────────────────────────────────────

func formatCxCell(v any) string {
	switch c := v.(type) {
	case nil:
		return "null"
	case bool:
		if c {
			return "true"
		}
		return "false"
	case int:
		return fmt.Sprintf("%d", c)
	case int64:
		return fmt.Sprintf("%d", c)
	case float64:
		return fmt.Sprintf("%g", c)
	case []any:
		parts := make([]string, len(c))
		for i, item := range c {
			parts[i] = formatCxCell(item)
		}
		return "[" + strings.Join(parts, ", ") + "]"
	case map[string]any:
		keys := make([]string, 0, len(c))
		for k := range c {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			parts = append(parts, formatCxKey(k)+": "+formatCxCell(c[k]))
		}
		return "{" + strings.Join(parts, ", ") + "}"
	case string:
		if strings.ContainsAny(c, " \t\n'[](){},") || c == "" {
			return "'" + strings.ReplaceAll(c, "'", "''") + "'"
		}
		return c
	default:
		return fmt.Sprintf("%v", c)
	}
}

func formatCxKey(k string) string {
	if k == "" {
		return "''"
	}
	for _, r := range k {
		if !((r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') ||
			(r >= '0' && r <= '9') || r == '_') {
			return "'" + strings.ReplaceAll(k, "'", "''") + "'"
		}
	}
	return k
}

func formatCsvCell(v any, delim byte) string {
	switch c := v.(type) {
	case nil:
		return ""
	case []any, map[string]any:
		bytes, _ := json.Marshal(c)
		return "\"" + strings.ReplaceAll(string(bytes), "\"", "\"\"") + "\""
	case bool:
		if c {
			return "true"
		}
		return "false"
	default:
		s := fmt.Sprintf("%v", c)
		if strings.ContainsRune(s, rune(delim)) || strings.ContainsAny(s, "\"\n\r") {
			return "\"" + strings.ReplaceAll(s, "\"", "\"\"") + "\""
		}
		return s
	}
}

// FromDataBinValue decodes CXDB payload bytes into a Go any value.
// Go's ToDataBin already strips the 4-byte LE size header before
// returning (see extractBinPayload), so the input here is the raw
// CXDB payload starting with the 'CXDB' magic bytes.
func FromDataBinValue(payload []byte) (any, error) {
	return decodeDataBin(payload)
}

File: lang/python/cxlib/table.py
"""CX Python binding — Public Table API per ADR 0018 D1.

Implements the 17-member canonical Table API surface against the V
core's `:table` blocks via the C ABI. Adopters import `cxlib.Table`
and call `Table.from_cx(src)` (or `Table.from_cx_all(src)` for
multi-table sources) to get a Table instance; the 17 methods provide
the public surface.

Per ADR 0018 §D2 per-binding naming: Python uses snake_case matching
the canonical surface exactly. `select` is renamed to `select_cols`
to avoid shadowing the standard-library `select` module.
"""

from __future__ import annotations

import json
from typing import Any, Iterator, Optional

from . import data_bin as _data_bin
from . import cx as _cx


class Table:
    """Immutable handle over a single `:table` block from a CX source.

    Per ADR 0018 §D3 tables are immutable values; modification returns
    a new Table. Cells admit any of the ADR 0017 §D5 Item kinds
    (scalars + Array / Map / Sequence). Arrays/maps materialize as
    Python `list` / `dict` respectively at the data-binding layer.
    """

    __slots__ = ("_cols", "_types", "_rows")

    def __init__(self, cols: list[str], types: list[str],
                 rows: list[list[Any]]):
        """Construct directly from cols / types / rows. Validates the
        4 invariants per ADR 0018 §D7. Adopters typically use
        `Table.from_cx(src)` instead.
        """
        if len(cols) != len(types):
            raise ValueError(
                f"cxlib: len(cols)={len(cols)} != len(types)={len(types)}"
            )
        seen: set[str] = set()
        for c in cols:
            if c in seen:
                raise ValueError(f'cxlib: duplicate column name "{c}"')
            seen.add(c)
        for row_idx, row in enumerate(rows):
            if len(row) != len(cols):
                raise ValueError(
                    f"cxlib: row {row_idx} has {len(row)} cells; "
                    f"expected {len(cols)}"
                )
        self._cols = list(cols)
        self._types = list(types)
        self._rows = [list(r) for r in rows]

    # ── Construction ─────────────────────────────────────────────────────────

    @classmethod
    def from_cx(cls, src: str) -> "Table":
        """Parse CX source and return the first `:table` block as a Table.

        Errors when the source contains no `:table` block; use
        `Table.from_cx_all(src)` for multi-table sources.
        """
        tables = cls.from_cx_all(src)
        if not tables:
            raise ValueError("cxlib: no :table block found in source")
        return tables[0]

    @classmethod
    def from_cx_all(cls, src: str) -> list["Table"]:
        """Parse CX source and return every `:table` block in document
        order. Walks the data_bin decode result recursively to find
        tables at any nesting depth.
        """
        # cx_to_data_bin produces FRAMED bytes (4-byte LE size + payload).
        # data_bin.decode handles the framing.
        framed = _cx.to_data_bin(src)
        decoded = _data_bin.decode(framed)
        tables: list[Table] = []
        _collect_tables(decoded, tables)
        return tables

    # ── Properties (4) ───────────────────────────────────────────────────────

    @property
    def cols(self) -> list[str]:
        """Ordered list of column names."""
        return list(self._cols)

    @property
    def types(self) -> list[str]:
        """Ordered list of column types (canonical string form, e.g.
        'int', 'string', 'bool'). Empty string means string default."""
        return list(self._types)

    @property
    def row_count(self) -> int:
        return len(self._rows)

    @property
    def col_count(self) -> int:
        return len(self._cols)

    # ── Access (9) ───────────────────────────────────────────────────────────

    def row(self, i: int) -> dict[str, Any]:
        """Row at index i as an ordered map."""
        if i < 0 or i >= len(self._rows):
            raise IndexError(
                f"cxlib: row index {i} out of bounds "
                f"[0, {len(self._rows)})"
            )
        return dict(zip(self._cols, self._rows[i]))

    def column(self, name: str) -> list[Any]:
        """All values in the named column, in row order."""
        try:
            col_idx = self._cols.index(name)
        except ValueError:
            raise KeyError(f'cxlib: unknown column "{name}"')
        return self.col_at(col_idx)

    def col_at(self, i: int) -> list[Any]:
        if i < 0 or i >= len(self._cols):
            raise IndexError(
                f"cxlib: column index {i} out of bounds "
                f"[0, {len(self._cols)})"
            )
        return [row[i] for row in self._rows]

    def cell(self, r: int, c: int) -> Any:
        if r < 0 or r >= len(self._rows):
            raise IndexError(
                f"cxlib: row index {r} out of bounds "
                f"[0, {len(self._rows)})"
            )
        if c < 0 or c >= len(self._cols):
            raise IndexError(
                f"cxlib: column index {c} out of bounds "
                f"[0, {len(self._cols)})"
            )
        return self._rows[r][c]

    def cell_by_name(self, r: int, name: str) -> Any:
        try:
            c = self._cols.index(name)
        except ValueError:
            raise KeyError(f'cxlib: unknown column "{name}"')
        return self.cell(r, c)

    def slice(self, start: int, end: int) -> "Table":
        if start < 0 or start > len(self._rows):
            raise IndexError(f"cxlib: slice start {start} out of bounds")
        if end < start or end > len(self._rows):
            raise IndexError(
                f"cxlib: slice end {end} out of bounds (start={start})"
            )
        new = Table.__new__(Table)
        new._cols = list(self._cols)
        new._types = list(self._types)
        new._rows = [list(r) for r in self._rows[start:end]]
        return new

    def head(self, n: int) -> "Table":
        return self.slice(0, max(0, min(n, len(self._rows))))

    def tail(self, n: int) -> "Table":
        start = max(0, len(self._rows) - n)
        return self.slice(start, len(self._rows))

    def select_cols(self, names: list[str]) -> "Table":
        """New Table with only the named columns, in the given order.

        Renamed from canonical `select` to avoid shadowing the
        standard-library `select` module name in Python.
        """
        indices: list[int] = []
        new_cols: list[str] = []
        new_types: list[str] = []
        for name in names:
            try:
                idx = self._cols.index(name)
            except ValueError:
                raise KeyError(f'cxlib: unknown column "{name}"')
            indices.append(idx)
            new_cols.append(self._cols[idx])
            new_types.append(self._types[idx])
        new_rows = [[row[i] for i in indices] for row in self._rows]
        new = Table.__new__(Table)
        new._cols = new_cols
        new._types = new_types
        new._rows = new_rows
        return new

    # ── Iteration (2) ────────────────────────────────────────────────────────

    def __iter__(self) -> Iterator[dict[str, Any]]:
        """`for row in t` yields ordered maps in row order."""
        for i in range(len(self._rows)):
            yield self.row(i)

    def iter_cols(self) -> Iterator[tuple[str, str, list[Any]]]:
        """Yields (name, type_name, values) per column in declaration
        order."""
        for i, name in enumerate(self._cols):
            yield (name, self._types[i], self.col_at(i))

    # ── Conversion (5) ───────────────────────────────────────────────────────

    def to_cx(self) -> str:
        """Canonical CX text — the `:table` block form per ADR 0018 §D6
        + spec/canonical.md §2.11.
        """
        header_parts = []
        for name, type_name in zip(self._cols, self._types):
            if type_name == "":
                header_parts.append(name)
            else:
                header_parts.append(f"{name}:{type_name}")
        header = " ".join(header_parts)
        lines = [f"[_ :table[{header}]"]
        for row in self._rows:
            cells = [_format_cx_cell(v) for v in row]
            lines.append("  " + " ".join(cells))
        lines.append("]")
        return "\n".join(lines) + "\n"

    def to_csv(self, delim: str = ",") -> str:
        """CSV (comma-separated) by default; pass `delim` for other
        single-char delimiters per ADR 0001 §D6. Collection cells emit
        as JSON-encoded strings per ADR 0001 §D7 / ADR 0018 §D6.
        """
        if len(delim) != 1:
            raise ValueError(
                f"cxlib: to_csv delim must be 1 char; got {len(delim)}"
            )
        out_lines = [delim.join(self._cols)]
        for row in self._rows:
            cells = [_format_csv_cell(v, delim) for v in row]
            out_lines.append(delim.join(cells))
        return "\r\n".join(out_lines) + "\r\n"

    def to_json(self) -> str:
        """Semantic JSON: list of row objects with cells as host-native
        JSON (Array → JSON array, Map → JSON object, scalars → JSON
        scalars). Per ADR 0018 §D6.
        """
        return json.dumps(self.to_dict_list(), separators=(",", ":"))

    def to_data_bin(self) -> bytes:
        """CXDB binary form via cx_to_data_bin. Plain `0x60` form for
        collection-cell tables (per Phase 2.2 wire-format rule);
        chunked `0x63` form is scalar-only and routed automatically.
        """
        return _cx.to_data_bin(self.to_cx())

    def to_dict_list(self) -> list[dict[str, Any]]:
        """Eager copy — each row materializes as a separate dict."""
        return [self.row(i) for i in range(len(self._rows))]

    # ── Equality ─────────────────────────────────────────────────────────────

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Table):
            return False
        return (
            self._cols == other._cols
            and self._types == other._types
            and self._rows == other._rows
        )

    def __repr__(self) -> str:
        return (
            f"Table(cols={self._cols!r}, types={self._types!r}, "
            f"rows={len(self._rows)})"
        )


# ── Internal: walk a decoded data_bin payload to find tables ─────────────────

def _collect_tables(value: Any, out: list[Table]) -> None:
    """Recursively walk a data_bin decode result, looking for tables.

    Python's data_bin decoder represents tables as `list[dict[str, Any]]`
    — list of row dicts where keys are column names. We detect tables
    by: a non-empty list of dicts where all dicts share the same key
    set (and that key set is non-empty). Nested structures recurse.

    Note: this is a heuristic; the data_bin format itself does not
    carry an explicit "this is a table" marker post-decode in the
    Python representation. For unambiguous table detection we'd need
    to extend the decoder to mark table values — followup item.
    """
    if isinstance(value, dict):
        for child in value.values():
            _collect_tables(child, out)
    elif isinstance(value, list):
        if _looks_like_table(value):
            cols = list(value[0].keys())
            # Types not preserved by data_bin decode; default to empty
            # string (string). Followup: extend decoder to carry types.
            types = [""] * len(cols)
            rows = [[row[c] for c in cols] for row in value]
            out.append(Table(cols=cols, types=types, rows=rows))
        else:
            for child in value:
                _collect_tables(child, out)


def _looks_like_table(value: list) -> bool:
    if not value:
        return False
    if not all(isinstance(item, dict) for item in value):
        return False
    keys = set(value[0].keys())
    if not keys:
        return False
    return all(set(item.keys()) == keys for item in value)


# ── Internal: cell formatters ─────────────────────────────────────────────────

def _format_cx_cell(v: Any) -> str:
    """Canonical CX cell rendering per ADR 0017 §D14."""
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return str(v)
    if isinstance(v, list):
        # Array literal
        return "[" + ", ".join(_format_cx_cell(item) for item in v) + "]"
    if isinstance(v, dict):
        # Map literal — keys lex-sorted per ADR 0017 §D14 canonical
        items = sorted(v.items())
        return "{" + ", ".join(
            f"{_format_cx_key(k)}: {_format_cx_cell(val)}"
            for k, val in items
        ) + "}"
    # String — quote if it contains whitespace / special chars
    s = str(v)
    if (not s) or any(c in s for c in (" \t\n'[](){},")):
        # Prefer single-quotes; escape internal single quotes with doubling.
        return "'" + s.replace("'", "''") + "'"
    return s


def _format_cx_key(k: Any) -> str:
    """Map-key rendering: bare-name when possible, quoted otherwise."""
    s = str(k)
    if s.isidentifier() and not any(c in s for c in "-:"):
        return s
    return "'" + s.replace("'", "''") + "'"


def _format_csv_cell(v: Any, delim: str) -> str:
    """CSV cell — collection cells emit as JSON-encoded strings per
    ADR 0001 §D7.
    """
    if v is None:
        return ""
    if isinstance(v, (list, dict)):
        # JSON-encode then quote per CSV rules.
        json_str = json.dumps(v, separators=(",", ":"))
        return '"' + json_str.replace('"', '""') + '"'
    if isinstance(v, bool):
        return "true" if v else "false"
    s = str(v)
    if delim in s or '"' in s or "\n" in s or "\r" in s:
        return '"' + s.replace('"', '""') + '"'
    return s
